### PR TITLE
Runnable with default config 

### PR DESCRIPTION
**Pull request by richtja**
Some runners (like instrumented) need config values for proper working.
When we are running tasks by `avocado run` or Job API the default
configuration is inserted to runners via runnable. The problem starts
when you run test runner with `runnable-run` command even a small test
as `avocado-runner runnable-run -k avocado-instrumented -u
examples/tests/passtest.py:PassTest.test` will fail because of missing
configuration. This commit solves this problem by adding default
configuration to the runnable.

Signed-off-by: Jan Richter <jarichte@redhat.com>

### CODE REVIEW

File: avocado/plugins/runner_nrunner.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def initialize(self):
                     'is in another host, or different port')
         settings.register_option(section=section,
                                  key='status_server_uri',
-                                 default='127.0.0.1:8888',
+                                 default=None,
```
Comment (beraldoleal) [inline avocado/plugins/runner_nrunner.py:77]: I can't see a problem with having this as default value. Can you explain better, what problem are you solving here, please?

**Comment by clebergnu** [comment]
Hi @richtja,

I understand your use case, but I'm afraid your proposal has some undesirable consequences.

When I came up with the design of the [runnables](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html#runnable), I intended to have a minimalist "description of an entity that can be executed and produce some kind of result".  That's one of the reasons that I'm trying to remove the non-essential information in https://github.com/avocado-framework/avocado/issues/5075 / https://github.com/avocado-framework/avocado/pull/5323.

One undesirable consequence of having all of Avocado's configuration in a runnable is that actual reproducibility of what was *intended* (instead of what was executed) is made much much harder.  One example may help to make the point across:

* I run `avocado-runner-asset runnable-run name='https://avocado-project.org/data/assets/SHA1SUM_JEOS23_XZ'` and get a failure.
* I open an issue, and copy the command line (or an equivalent recipe file)

Now, if all of Avocado's configuration would be considered part of the runnable, someone would have a hard time excluding configuration from the possible causes of failure.  It would be a tiring trial and error process.

Think of the amount of system specific data would be pushed into the reproducible, for instance, the location where the file would be saved/cached (`/home/cleber/avocado/data/cache` in my case, but most likely not on the reproducer system).

In summary, to be able to automatically run "reproducibles", we need to know what constitutes the basic instructions on what to run (the purpose of a runnable), and the configuration which can be system specific.  I believe that in most cases, users will want to run reproducibles using their own system configuration (to be able to distinguish between the behavior with different configurations), but another utility or feature can be added to optionally inject the configuration used previously in another execution -- as long as it's *configuration* nature is clear.

**Comment by richtja** [comment]
> Hi @richtja,
> 
> I understand your use case, but I'm afraid your proposal has some undesirable consequences.
> 
> When I came up with the design of the [runnables](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html#runnable), I intended to have a minimalist "description of an entity that can be executed and produce some kind of result". That's one of the reasons that I'm trying to remove the non-essential information in #5075 / #5323.
> 

Hi @clebergnu a went throw your work in #5323, and I agree that it almost solves my problem, and it is better because it uses only necessary configuration for runners. What I am missing there,  is the same functionality which [nrunner plugin](https://github.com/avocado-framework/avocado/pull/5323/commits/4a227e390af25bfa7c594a910a235660be8c6edc) has, added also to the `avocado-runner task-run` and others.

> One undesirable consequence of having all of Avocado's configuration in a runnable is that actual reproducibility of what was _intended_ (instead of what was executed) is made much much harder. One example may help to make the point across:
> 
>     * I run `avocado-runner-asset runnable-run name='https://avocado-project.org/data/assets/SHA1SUM_JEOS23_XZ'` and get a failure.
> 
>     * I open an issue, and copy the command line (or an equivalent recipe file)
> 
> 
> Now, if all of Avocado's configuration would be considered part of the runnable, someone would have a hard time excluding configuration from the possible causes of failure. It would be a tiring trial and error process.
> 
> Think of the amount of system specific data would be pushed into the reproducible, for instance, the location where the file would be saved/cached (`/home/cleber/avocado/data/cache` in my case, but most likely not on the reproducer system).
> 
> In summary, to be able to automatically run "reproducibles", we need to know what constitutes the basic instructions on what to run (the purpose of a runnable), and the configuration which can be system specific. I believe that in most cases, users will want to run reproducibles using their own system configuration (to be able to distinguish between the behavior with different configurations), but another utility or feature can be added to optionally inject the configuration used previously in another execution -- as long as its _configuration_ nature is clear.


I think my changes doesn't make reproducibility worse. I only add default configuration (which is base on the system where the runnable is run) if the configuration is not specified by the user.  IMO, we have this problem with reproducibility anyway because the `avocado run` command adds the whole configuration to the runnable anyway.

**Comment by richtja** [comment]
Because of https://github.com/avocado-framework/avocado/pull/5323 and its usage for this problem, I am changing this to draft for now.

**Comment by clebergnu** [comment]
> > Hi @richtja,
> > I understand your use case, but I'm afraid your proposal has some undesirable consequences.
> > When I came up with the design of the [runnables](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html#runnable), I intended to have a minimalist "description of an entity that can be executed and produce some kind of result". That's one of the reasons that I'm trying to remove the non-essential information in #5075 / #5323.
> 
> Hi @clebergnu a went throw your work in #5323, and I agree that it almost solves my problem, and it is better because it uses only necessary configuration for runners. What I am missing there, is the same functionality which [nrunner plugin](https://github.com/avocado-framework/avocado/pull/5323/commits/4a227e390af25bfa7c594a910a235660be8c6edc) has, added also to the `avocado-runner task-run` and others.
> 

Let me try to properly understand your use case.  You want to be able to run `avocado-runner task-run` and have, by default, "essential configuration" added to it, right?

If so, I can pretty much relate to that use case.  One of the high points IMO of the current architecture is that it is (or should be) easy to debug the behavior by re-running an `avocado-runner*` command.  But then the system/execution specific configuration that was added, say from the main Avocado configuration or from dynamic values such as the status server location, will be missing.  Just as a side note, this is not a problem specific to Avocado, and I can relate to the same difficulty when trying to reproduce say QEMU command lines (the location of temporary files, sockets, etc, will be different).

I see one way of solving this:

1) properly handling the two different types of input
2) optionally adding/filtering some of them out

The example I can come up with is when `avocado run /bin/true` produces:

```
avocado-runner-exec-test task-run -i 1-/bin/true -j 263cbe79efdd79e21509b458970050571fea3be0 -k exec-test -u /bin/true -c {"run.test_parameters": [], "runner.exectest.exitcodes.skip": []} variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]} output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true -s /var/tmp/avocado_i5t1xtnt/.status_server.sock
```

Let's see if we are able to distinguish between what is "description of what to run" and what is system/run-time information:

1) `-i 1-/bin/true`: probably irrelevant for reproducibility purposes, unless a job is being replayed and the exact same result output structure is expected from the replayed job.  IMO this is correctly categorized as a *Task* parameter, because it does *not* describe what is to be run.

2) `-k exec-test`: this is relevant because it will determine how "everything else" (like the `uri`) will be run (in this case a process will be executed).

3) `-u /bin/true`: quite obviously relevant for reproducibility

4) `"run.test_parameters": [],` relevant for reproducibility purposes, but *wrongly* relying on "configuration" data.  IMO this should be a clearly defined input for the runnable.  Originally, the idea was to *only* use `-a` and `KEY=VAL` for data such as this, but due to the legacy code we failed to map those properly.  IMO there's an action item here.

5) `"runner.exectest.exitcodes.skip": []`: this is a bit of a grey area, but if I had to choose, I'll say this should be part of the runnable description.  How the runner will respond to the exit status process execution of `uri` is IMO an integral part of the `exec-test` runner responsibility.  BTW, we could have a *different* setting in Avocado that transforms results (in general, despite of their kinds) as an extra step.  It could be used for waiving of known existing failures, etc...

6) `variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}`: this is relevant for reproducibility purposes.  It's unfortunate that our variants/parameter syntax/data is so convoluted, but that's because of its features (paths, IDs), and a matter for a different discussion.

7) `output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true`: this is *not* relevant for reproducibility purposes. IMO this should not be a runnable `KEY=VAL` value.

8) `-s /var/tmp/avocado_i5t1xtnt/.status_server.sock`: not relevant for reproducibility purposes. IMO this is correctly categorized as a *Task* parameter, because it does *not* describe what is to be run.

If we are able to clearly determine what is and is not relevant for reproducibility purposes, we may end up with a command like such as:

```
$ avocado-runner-exec-test task-run -i 1-/bin/true -j 263cbe79efdd79e21509b458970050571fea3be0 -k exec-test -u /bin/true --output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true -s /var/tmp/avocado_i5t1xtnt/.status_server.sock exectest.exitcodes.skip=1,2 variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]} 
```

And with that, we could have utility such as:

```
$ avocado-reproduce-runnable avocado-runner-exec-test task-run -i 1-/bin/true -j 263cbe79efdd79e21509b458970050571fea3be0 -k exec-test -u /bin/true --output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true -s /var/tmp/avocado_i5t1xtnt/.status_server.sock exectest.exitcodes.skip=1,2 variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}

avocado-runner-exec-test task-run -k exec-test -u /bin/true exectest.exitcodes.skip=1,2 variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}
```

So users can copy and paste the "sanitized" version of the command Avocado run.  Another possibility, which may be even easier to handle, is to rely more on recipe files then on the command line arguments.

Sorry for the long response, but hopefully this discussion will lead us to a better solution.

> > One undesirable consequence of having all of Avocado's configuration in a runnable is that actual reproducibility of what was _intended_ (instead of what was executed) is made much much harder. One example may help to make the point across:
> > ```
> > * I run `avocado-runner-asset runnable-run name='https://avocado-project.org/data/assets/SHA1SUM_JEOS23_XZ'` and get a failure.
> > 
> > * I open an issue, and copy the command line (or an equivalent recipe file)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Now, if all of Avocado's configuration would be considered part of the runnable, someone would have a hard time excluding configuration from the possible causes of failure. It would be a tiring trial and error process.
> > Think of the amount of system specific data would be pushed into the reproducible, for instance, the location where the file would be saved/cached (`/home/cleber/avocado/data/cache` in my case, but most likely not on the reproducer system).
> > In summary, to be able to automatically run "reproducibles", we need to know what constitutes the basic instructions on what to run (the purpose of a runnable), and the configuration which can be system specific. I believe that in most cases, users will want to run reproducibles using their own system configuration (to be able to distinguish between the behavior with different configurations), but another utility or feature can be added to optionally inject the configuration used previously in another execution -- as long as its _configuration_ nature is clear.
> 
> I think my changes doesn't make reproducibility worse. I only add default configuration (which is base on the system where the runnable is run) if the configuration is not specified by the user. IMO, we have this problem with reproducibility anyway because the `avocado run` command adds the whole configuration to the runnable anyway.

What I mean is that, because an `avocado-runner*` execution currently does not know how to get configuration from the system, a change in the system configuration will not "invisibly" make its way into its behavior while executing.  

I agree that this is an uncommon use case (few people will run `avocado-runner-*`) but  when comparing the execution of `avocado-runner-*` before and after this proposed change, IMO there could be more variables (all local configuration) being pushed to `avocado-runner`.

**Comment by richtja** [comment]
> > > Hi @richtja,
> > > I understand your use case, but I'm afraid your proposal has some undesirable consequences.
> > > When I came up with the design of the [runnables](https://avocado-framework.readthedocs.io/en/96.0/guides/contributor/chapters/runners.html#runnable), I intended to have a minimalist "description of an entity that can be executed and produce some kind of result". That's one of the reasons that I'm trying to remove the non-essential information in #5075 / #5323.
> > 
> > 
> > Hi @clebergnu a went throw your work in #5323, and I agree that it almost solves my problem, and it is better because it uses only necessary configuration for runners. What I am missing there, is the same functionality which [nrunner plugin](https://github.com/avocado-framework/avocado/pull/5323/commits/4a227e390af25bfa7c594a910a235660be8c6edc) has, added also to the `avocado-runner task-run` and others.
> 
> Let me try to properly understand your use case. You want to be able to run `avocado-runner task-run` and have, by default, "essential configuration" added to it, right?
> 
> If so, I can pretty much relate to that use case. One of the high points IMO of the current architecture is that it is (or should be) easy to debug the behavior by re-running an `avocado-runner*` command. But then the system/execution specific configuration that was added, say from the main Avocado configuration or from dynamic values such as the status server location, will be missing. Just as a side note, this is not a problem specific to Avocado, and I can relate to the same difficulty when trying to reproduce say QEMU command lines (the location of temporary files, sockets, etc, will be different).
> 
> I see one way of solving this:
> 
>     1. properly handling the two different types of input
> 
>     2. optionally adding/filtering some of them out
> 
> 
> The example I can come up with is when `avocado run /bin/true` produces:
> 
> ```
> avocado-runner-exec-test task-run -i 1-/bin/true -j 263cbe79efdd79e21509b458970050571fea3be0 -k exec-test -u /bin/true -c {"run.test_parameters": [], "runner.exectest.exitcodes.skip": []} variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]} output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true -s /var/tmp/avocado_i5t1xtnt/.status_server.sock
> ```
> 
> Let's see if we are able to distinguish between what is "description of what to run" and what is system/run-time information:
> 
>     1. `-i 1-/bin/true`: probably irrelevant for reproducibility purposes, unless a job is being replayed and the exact same result output structure is expected from the replayed job.  IMO this is correctly categorized as a _Task_ parameter, because it does _not_ describe what is to be run.
> 
>     2. `-k exec-test`: this is relevant because it will determine how "everything else" (like the `uri`) will be run (in this case a process will be executed).
> 
>     3. `-u /bin/true`: quite obviously relevant for reproducibility
> 
>     4. `"run.test_parameters": [],` relevant for reproducibility purposes, but _wrongly_ relying on "configuration" data.  IMO this should be a clearly defined input for the runnable.  Originally, the idea was to _only_ use `-a` and `KEY=VAL` for data such as this, but due to the legacy code we failed to map those properly.  IMO there's an action item here.
> 
>     5. `"runner.exectest.exitcodes.skip": []`: this is a bit of a grey area, but if I had to choose, I'll say this should be part of the runnable description.  How the runner will respond to the exit status process execution of `uri` is IMO an integral part of the `exec-test` runner responsibility.  BTW, we could have a _different_ setting in Avocado that transforms results (in general, despite of their kinds) as an extra step.  It could be used for waiving of known existing failures, etc...
> 
>     6. `variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}`: this is relevant for reproducibility purposes.  It's unfortunate that our variants/parameter syntax/data is so convoluted, but that's because of its features (paths, IDs), and a matter for a different discussion.
> 
>     7. `output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true`: this is _not_ relevant for reproducibility purposes. IMO this should not be a runnable `KEY=VAL` value.
> 
>     8. `-s /var/tmp/avocado_i5t1xtnt/.status_server.sock`: not relevant for reproducibility purposes. IMO this is correctly categorized as a _Task_ parameter, because it does _not_ describe what is to be run.
> 

I think your example is great for explaining the problem which I am trying to solve and also shows the possible solution. We can basically split the parameters into three groups:

1. Values which defines the test and has to be specified by user:
* `-k exec-test`
*  `-u /bin/true`
2. Values created by Avocado and not relevant for test itself:
* `-i 1-/bin/true`
* `output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true`
* `-s /var/tmp/avocado_i5t1xtnt/.status_server.sock`
3. Values which changes test behavior (they are relevant for reproducibility), but it wouldn't break the runner if they are not set:
* `variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}`
4.  Values which changes test behavior (they are relevant for reproducibility), and it will break the runner if they are not set:
* `"runner.exectest.exitcodes.skip": []`

The group 4 is basically essential config variables for each runner. Right now when you want to run `avocado-runner task-run` you have to specify those variables even thou you want to use default values. If you don't do that, you will break the runner. IMO shouldn't be necessary fill the default values, because they are default. I think with #5326 the solution for this problem is much easier, because those config values are defined by `CONFIGURATION_USED` list. So we can add only default values of `CONFIGURATION_USED` into the runnable and the problem will be solved.

> 
> If we are able to clearly determine what is and is not relevant for reproducibility purposes, we may end up with a command like such as:
> 
> ```
> $ avocado-runner-exec-test task-run -i 1-/bin/true -j 263cbe79efdd79e21509b458970050571fea3be0 -k exec-test -u /bin/true --output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true -s /var/tmp/avocado_i5t1xtnt/.status_server.sock exectest.exitcodes.skip=1,2 variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]} 
> ```
> 
> And with that, we could have utility such as:
> 
> ```
> $ avocado-reproduce-runnable avocado-runner-exec-test task-run -i 1-/bin/true -j 263cbe79efdd79e21509b458970050571fea3be0 -k exec-test -u /bin/true --output_dir=/home/cleber/avocado/job-results/job-2022-04-12T10.15-263cbe7/test-results/1-_bin_true -s /var/tmp/avocado_i5t1xtnt/.status_server.sock exectest.exitcodes.skip=1,2 variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}
> 
> avocado-runner-exec-test task-run -k exec-test -u /bin/true exectest.exitcodes.skip=1,2 variant=json:{"paths": ["/run/*"], "variant_id": null, "variant": [["/", []]]}
> ```
> 
> So users can copy and paste the "sanitized" version of the command Avocado run. Another possibility, which may be even easier to handle, is to rely more on recipe files then on the command line arguments.
> 
> Sorry for the long response, but hopefully this discussion will lead us to a better solution.
> 

> > > One undesirable consequence of having all of Avocado's configuration in a runnable is that actual reproducibility of what was _intended_ (instead of what was executed) is made much much harder. One example may help to make the point across:
> > > ```
> > > * I run `avocado-runner-asset runnable-run name='https://avocado-project.org/data/assets/SHA1SUM_JEOS23_XZ'` and get a failure.
> > > 
> > > * I open an issue, and copy the command line (or an equivalent recipe file)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Now, if all of Avocado's configuration would be considered part of the runnable, someone would have a hard time excluding configuration from the possible causes of failure. It would be a tiring trial and error process.
> > > Think of the amount of system specific data would be pushed into the reproducible, for instance, the location where the file would be saved/cached (`/home/cleber/avocado/data/cache` in my case, but most likely not on the reproducer system).
> > > In summary, to be able to automatically run "reproducibles", we need to know what constitutes the basic instructions on what to run (the purpose of a runnable), and the configuration which can be system specific. I believe that in most cases, users will want to run reproducibles using their own system configuration (to be able to distinguish between the behavior with different configurations), but another utility or feature can be added to optionally inject the configuration used previously in another execution -- as long as its _configuration_ nature is clear.
> > 
> > 
> > I think my changes doesn't make reproducibility worse. I only add default configuration (which is base on the system where the runnable is run) if the configuration is not specified by the user. IMO, we have this problem with reproducibility anyway because the `avocado run` command adds the whole configuration to the runnable anyway.
> 
> What I mean is that, because an `avocado-runner*` execution currently does not know how to get configuration from the system, a change in the system configuration will not "invisibly" make its way into its behavior while executing.
> 
> I agree that this is an uncommon use case (few people will run `avocado-runner-*`) but when comparing the execution of `avocado-runner-*` before and after this proposed change, IMO there could be more variables (all local configuration) being pushed to `avocado-runner`.

**Comment by richtja** [comment]
Because the https://github.com/avocado-framework/avocado/pull/5326 was merged, I created a new version which uses those changes.

**Comment by richtja** [comment]
Closing in favor of #5363
